{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2086f911-0d14-4f14-b2f8-e91d9e1db8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "# def load_file(path, chunk_size=100_000):\n",
    "#     chunks = []\n",
    "#     for chunk in pd.read_csv(path, chunksize=chunk_size):\n",
    "#         chunks.append(chunk)\n",
    "#     return pd.concat(chunks, ignore_index=True)\n",
    "\n",
    "\n",
    "# df_lof_final_result = load_file(\"C:/Users/maliz/thesa/UKbiobank/python_and_R/pipeline/result/lof_final_result.csv\")\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af5fd829-fde6-48b5-b004-1ed772718742",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_missense_paralog = pd.read_csv(\"paralogs/missense_paralogs_results.csv\")\n",
    "df_missense_paralog = df_missense_paralog.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b39a4c38-3818-432e-b681-0e4c5e4a0d51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected 62.24557962048208\n",
      "observed 62.0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "missense_expected = df_missense_paralog.expected_both_missense.sum()\n",
    "missense_observed = df_missense_paralog.both_missense.sum()\n",
    "\n",
    "print (\"expected \" +str(df_missense_paralog.expected_both_missense.sum()))\n",
    "print (\"observed \" +str(df_missense_paralog.both_missense.sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "24cfc59d-9aad-4c04-ac51-7828708c1588",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lof_paralogs_result = pd.read_csv(\"paralogs/lof_paralogs_results.csv\")\n",
    "df_lof_paralogs_result = df_lof_paralogs_result.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "314ffb0b-5b00-4214-a136-48141aaec251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(159, 8)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_lof_paralogs_result.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9af4f506-dc98-4db2-b46f-80626d9577b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected 5.007291480633059\n",
      "observed 3.0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "lof_expected = df_lof_paralogs_result.expected_both_lof.sum()\n",
    "lof_observed = df_lof_paralogs_result.both_lof.sum()\n",
    "\n",
    "\n",
    "print (\"expected \" +str(df_lof_paralogs_result.expected_both_lof.sum()))\n",
    "print (\"observed \" +str(df_lof_paralogs_result.both_lof.sum()))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3e0e8f14-fc55-4a85-8603-fe72ba5a404b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_missense_lof_paralog = pd.read_csv(\"paralogs/missense_lof_paralogs_results.csv\")\n",
    "df_missense_lof_paralog = df_missense_lof_paralog.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "419aecd1-7eb3-4a71-bf7b-06fda7e1e0c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected 16.18177649957329\n",
      "observed 11.0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "missense_lof_expected = df_missense_lof_paralog.expected_both_missense_lof.sum()\n",
    "missense_lof_observed = df_missense_lof_paralog.both_missense_lof.sum()\n",
    "\n",
    "print (\"expected \" +str(df_missense_lof_paralog.expected_both_missense_lof.sum()))\n",
    "print (\"observed \" +str(df_missense_lof_paralog.both_missense_lof.sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "044b888f-b904-4e89-901f-3d886aadfb9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lof_missense_paralog = pd.read_csv(\"paralogs/lof_missense_paralogs_results.csv\")\n",
    "df_lof_missense_paralog = df_lof_missense_paralog.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "392f5c52-dc2a-4acd-a5dc-52863c7c54a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected 12.665882690856895\n",
      "observed 9.0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "lof_missense_expected = df_lof_missense_paralog.expected_both_lof_missense.sum()\n",
    "lof_missense_observed = df_lof_missense_paralog.both_lof_missense.sum()\n",
    "\n",
    "print (\"expected \" +str(df_lof_missense_paralog.expected_both_lof_missense.sum()))\n",
    "print (\"observed \" +str(df_lof_missense_paralog.both_lof_missense.sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c65be142-21d1-47d0-a74c-17448e6c771e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5212838896306673"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import poisson\n",
    "\n",
    "p_value_missense = poisson.cdf(missense_observed, missense_expected)\n",
    "p_value_missense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "07dd055d-8ac7-455c-875c-6da0f9ee23d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.264003874004754"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import poisson\n",
    "\n",
    "p_value_lof = poisson.cdf(lof_observed,lof_expected)\n",
    "p_value_lof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "13ec7978-d3b6-4237-be3e-e93d4549fe2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.18903207639021352"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import poisson\n",
    "\n",
    "p_value_lof_missense = poisson.cdf(lof_missense_observed,lof_missense_expected)\n",
    "p_value_lof_missense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5b411f19-55dd-4408-944e-c1814e42c694",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.11823047956674507"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import poisson\n",
    "\n",
    "p_value_missense_lof = poisson.cdf(missense_lof_observed,missense_lof_expected)\n",
    "p_value_missense_lof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8c20f9e4-733d-40cf-bd1d-5459f925a0b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "paralogs_high_corr = pd.read_csv(\"paralogs_high_corr.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "05225860-4bf5-4ab6-af68-ced0444aff25",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "28782ae3-6ffa-4917-8452-1a9ef929e6de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol_1</th>\n",
       "      <th>symbol_2</th>\n",
       "      <th>Paralogs names</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>chr_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>chr_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADSS</td>\n",
       "      <td>ADSSL1</td>\n",
       "      <td>ADSSL1</td>\n",
       "      <td>0.256821</td>\n",
       "      <td>ENSG00000267312</td>\n",
       "      <td>17</td>\n",
       "      <td>ENSG00000185100</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H2AFV</td>\n",
       "      <td>H2AFZ</td>\n",
       "      <td>H2AFV</td>\n",
       "      <td>0.116486</td>\n",
       "      <td>ENSG00000105968</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSG00000164032</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARF1</td>\n",
       "      <td>ARF3</td>\n",
       "      <td>ARF3, ARF6, ARF4, ARL17B, ARF5</td>\n",
       "      <td>0.177112</td>\n",
       "      <td>ENSG00000143761</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000134287</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARF1</td>\n",
       "      <td>ARF4</td>\n",
       "      <td>ARF3, ARF6, ARF4, ARL17B, ARF5</td>\n",
       "      <td>0.127893</td>\n",
       "      <td>ENSG00000143761</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000168374</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARF1</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>ARF3, ARF6, ARF4, ARL17B, ARF5</td>\n",
       "      <td>0.142654</td>\n",
       "      <td>ENSG00000143761</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  symbol_1 symbol_2                  Paralogs names  Correlation  \\\n",
       "0     ADSS   ADSSL1                          ADSSL1     0.256821   \n",
       "1    H2AFV    H2AFZ                           H2AFV     0.116486   \n",
       "2     ARF1     ARF3  ARF3, ARF6, ARF4, ARL17B, ARF5     0.177112   \n",
       "3     ARF1     ARF4  ARF3, ARF6, ARF4, ARL17B, ARF5     0.127893   \n",
       "4     ARF1     ARF5  ARF3, ARF6, ARF4, ARL17B, ARF5     0.142654   \n",
       "\n",
       "            gene_1  chr_1           gene_2  chr_2  \n",
       "0  ENSG00000267312     17  ENSG00000185100     14  \n",
       "1  ENSG00000105968      7  ENSG00000164032      4  \n",
       "2  ENSG00000143761      1  ENSG00000134287     12  \n",
       "3  ENSG00000143761      1  ENSG00000168374      3  \n",
       "4  ENSG00000143761      1  ENSG00000004059      7  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f4827c-95b2-4a6f-ad05-2931f4ade441",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_missense_lof_paralog.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7583390c-9a71-4b66-84ad-f6a74a506889",
   "metadata": {},
   "outputs": [],
   "source": [
    "ENSG00000267312 (17) , ENSG00000105968 (7)  , ENSG00000143761(1) , ENSG00000185100 (14) , ENSG00000164032(4) , \n",
    "ENSG00000134287(12) , ENSG00000168374(3) , ENSG00000004059 (7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "df657789-622d-4922-a7b6-9e1d0e420061",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "def load_file(path, chunk_size=100_000):\n",
    "    chunks = []\n",
    "    for chunk in pd.read_csv(path, chunksize=chunk_size):\n",
    "        chunks.append(chunk)\n",
    "    return pd.concat(chunks, ignore_index=True)\n",
    "\n",
    "\n",
    "df_lof_final_result = load_file(\"C:/Users/maliz/thesa/UKbiobank/python_and_R/pipeline/result/lof_final_result.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3f1c5104-3ecc-4484-8429-9f143ed13f1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>both_lof</th>\n",
       "      <th>lof_1</th>\n",
       "      <th>lof_2</th>\n",
       "      <th>expected_both_lof</th>\n",
       "      <th>p-value_lof</th>\n",
       "      <th>z-score_lof</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSG00000008735</td>\n",
       "      <td>0.0</td>\n",
       "      <td>132</td>\n",
       "      <td>70</td>\n",
       "      <td>0.027024</td>\n",
       "      <td>0.973338</td>\n",
       "      <td>-1.932290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSG00000015475</td>\n",
       "      <td>0.0</td>\n",
       "      <td>132</td>\n",
       "      <td>80</td>\n",
       "      <td>0.030884</td>\n",
       "      <td>0.969588</td>\n",
       "      <td>-1.874770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSG00000025708</td>\n",
       "      <td>0.0</td>\n",
       "      <td>132</td>\n",
       "      <td>476</td>\n",
       "      <td>0.183761</td>\n",
       "      <td>0.832134</td>\n",
       "      <td>-0.962634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSG00000025770</td>\n",
       "      <td>0.0</td>\n",
       "      <td>132</td>\n",
       "      <td>171</td>\n",
       "      <td>0.066015</td>\n",
       "      <td>0.936117</td>\n",
       "      <td>-1.522969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSG00000040608</td>\n",
       "      <td>0.0</td>\n",
       "      <td>132</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003088</td>\n",
       "      <td>0.996916</td>\n",
       "      <td>-2.738750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135893222</th>\n",
       "      <td>ENSG00000288558</td>\n",
       "      <td>ENSG00000286135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.999912</td>\n",
       "      <td>-3.751941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135893223</th>\n",
       "      <td>ENSG00000288558</td>\n",
       "      <td>ENSG00000286140</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>-4.294006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135893224</th>\n",
       "      <td>ENSG00000288558</td>\n",
       "      <td>ENSG00000286190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>142</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.999585</td>\n",
       "      <td>-3.342449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135893225</th>\n",
       "      <td>ENSG00000288558</td>\n",
       "      <td>ENSG00000286224</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.999953</td>\n",
       "      <td>-3.906645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135893226</th>\n",
       "      <td>ENSG00000288558</td>\n",
       "      <td>ENSG00000286235</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>135893227 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_1           gene_2  both_lof  lof_1  lof_2  \\\n",
       "0          ENSG00000000419  ENSG00000008735       0.0    132     70   \n",
       "1          ENSG00000000419  ENSG00000015475       0.0    132     80   \n",
       "2          ENSG00000000419  ENSG00000025708       0.0    132    476   \n",
       "3          ENSG00000000419  ENSG00000025770       0.0    132    171   \n",
       "4          ENSG00000000419  ENSG00000040608       0.0    132      8   \n",
       "...                    ...              ...       ...    ...    ...   \n",
       "135893222  ENSG00000288558  ENSG00000286135       0.0      1     30   \n",
       "135893223  ENSG00000288558  ENSG00000286140       0.0      1      3   \n",
       "135893224  ENSG00000288558  ENSG00000286190       0.0      1    142   \n",
       "135893225  ENSG00000288558  ENSG00000286224       0.0      1     16   \n",
       "135893226  ENSG00000288558  ENSG00000286235       0.0      1      0   \n",
       "\n",
       "           expected_both_lof  p-value_lof  z-score_lof  \n",
       "0                   0.027024     0.973338    -1.932290  \n",
       "1                   0.030884     0.969588    -1.874770  \n",
       "2                   0.183761     0.832134    -0.962634  \n",
       "3                   0.066015     0.936117    -1.522969  \n",
       "4                   0.003088     0.996916    -2.738750  \n",
       "...                      ...          ...          ...  \n",
       "135893222           0.000088     0.999912    -3.751941  \n",
       "135893223           0.000009     0.999991    -4.294006  \n",
       "135893224           0.000415     0.999585    -3.342449  \n",
       "135893225           0.000047     0.999953    -3.906645  \n",
       "135893226           0.000000     1.000000         -inf  \n",
       "\n",
       "[135893227 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_lof_final_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "df39eb77-4184-4104-bfba-e9d1c20f9966",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>both_lof</th>\n",
       "      <th>lof_1</th>\n",
       "      <th>lof_2</th>\n",
       "      <th>expected_both_lof</th>\n",
       "      <th>p-value_lof</th>\n",
       "      <th>z-score_lof</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_1, gene_2, both_lof, lof_1, lof_2, expected_both_lof, p-value_lof, z-score_lof]\n",
       "Index: []"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_lof_final_result[df_lof_final_result.gene_1 == 'ENSG00000267312']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2875fdb8-17fd-4913-8042-af2c12cb6e9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>both_lof</th>\n",
       "      <th>lof_1</th>\n",
       "      <th>lof_2</th>\n",
       "      <th>expected_both_lof</th>\n",
       "      <th>p-value_lof</th>\n",
       "      <th>z-score_lof</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_1, gene_2, both_lof, lof_1, lof_2, expected_both_lof, p-value_lof, z-score_lof]\n",
       "Index: []"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_lof_final_result[df_lof_final_result.gene_2 == 'ENSG00000267312']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9b4e5040-2cc6-4c8f-b6b1-f062ddb52228",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Set the folder path\n",
    "folder_path = \"C:/Users/maliz/thesa/UKbiobank/python_and_R/pipeline/chr17\"\n",
    "\n",
    "# List to store DataFrames\n",
    "dfs = []\n",
    "\n",
    "# Loop over all files in the folder\n",
    "for filename in os.listdir(folder_path):\n",
    "    if filename.endswith(\".csv\"):\n",
    "        file_path = os.path.join(folder_path, filename)\n",
    "        df = pd.read_csv(file_path)\n",
    "        dfs.append(df)\n",
    "\n",
    "# Example: concatenate all dataframes into one\n",
    "df_combined = pd.concat(dfs, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ba803fc9-478c-420b-b1c2-66a97b20ec96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNV_separated</th>\n",
       "      <th>transcript</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>mut</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17_1001015_G_A</td>\n",
       "      <td>ENST00000327158.5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Gly171Asp</td>\n",
       "      <td>ENSG00000177370</td>\n",
       "      <td>TIMM22</td>\n",
       "      <td>17.0</td>\n",
       "      <td>997129.0</td>\n",
       "      <td>1003671.0</td>\n",
       "      <td>17:997129-1003671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17_1001026_G_C</td>\n",
       "      <td>ENST00000327158.5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Gly175Arg</td>\n",
       "      <td>ENSG00000177370</td>\n",
       "      <td>TIMM22</td>\n",
       "      <td>17.0</td>\n",
       "      <td>997129.0</td>\n",
       "      <td>1003671.0</td>\n",
       "      <td>17:997129-1003671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17_1001026_G_A</td>\n",
       "      <td>ENST00000327158.5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Gly175Arg</td>\n",
       "      <td>ENSG00000177370</td>\n",
       "      <td>TIMM22</td>\n",
       "      <td>17.0</td>\n",
       "      <td>997129.0</td>\n",
       "      <td>1003671.0</td>\n",
       "      <td>17:997129-1003671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17_1001039_G_A</td>\n",
       "      <td>ENST00000327158.5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Cys179Tyr</td>\n",
       "      <td>ENSG00000177370</td>\n",
       "      <td>TIMM22</td>\n",
       "      <td>17.0</td>\n",
       "      <td>997129.0</td>\n",
       "      <td>1003671.0</td>\n",
       "      <td>17:997129-1003671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17_1001047_T_G</td>\n",
       "      <td>ENST00000327158.5</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Phe182Val</td>\n",
       "      <td>ENSG00000177370</td>\n",
       "      <td>TIMM22</td>\n",
       "      <td>17.0</td>\n",
       "      <td>997129.0</td>\n",
       "      <td>1003671.0</td>\n",
       "      <td>17:997129-1003671</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SNV_separated         transcript     mutation_type          mut  \\\n",
       "0  chr17_1001015_G_A  ENST00000327158.5  missense_variant  p.Gly171Asp   \n",
       "1  chr17_1001026_G_C  ENST00000327158.5  missense_variant  p.Gly175Arg   \n",
       "2  chr17_1001026_G_A  ENST00000327158.5  missense_variant  p.Gly175Arg   \n",
       "3  chr17_1001039_G_A  ENST00000327158.5  missense_variant  p.Cys179Tyr   \n",
       "4  chr17_1001047_T_G  ENST00000327158.5  missense_variant  p.Phe182Val   \n",
       "\n",
       "              gene  symbol  chromosome     start        end           location  \n",
       "0  ENSG00000177370  TIMM22        17.0  997129.0  1003671.0  17:997129-1003671  \n",
       "1  ENSG00000177370  TIMM22        17.0  997129.0  1003671.0  17:997129-1003671  \n",
       "2  ENSG00000177370  TIMM22        17.0  997129.0  1003671.0  17:997129-1003671  \n",
       "3  ENSG00000177370  TIMM22        17.0  997129.0  1003671.0  17:997129-1003671  \n",
       "4  ENSG00000177370  TIMM22        17.0  997129.0  1003671.0  17:997129-1003671  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b3f65c1b-1973-4368-a345-ad655f31432e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNV_separated</th>\n",
       "      <th>transcript</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>mut</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SNV_separated, transcript, mutation_type, mut, gene, symbol, chromosome, start, end, location]\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combined[df_combined.gene == 'ENSG00000267312']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4cd27f0-1dd5-435c-9ab8-0f1f559e3a24",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
