{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d46fff63-f519-41c0-9e15-a948af0d5fc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done! Check gene_paralog_pairs_with_original.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from itertools import combinations\n",
    "\n",
    "df = pd.read_csv('iyac134_supplemental_table_s5.xlsx - Pralogs Data.csv')\n",
    "df  =df[df[\"Suggested mechanism\"] == 'redundancy']\n",
    "output_rows = []\n",
    "\n",
    "for _, row in df.iterrows():\n",
    "    main_gene = row['Tissue-specific gene'].strip()\n",
    "    paralogs = [p.strip() for p in row['Paralogs names'].split(',')]\n",
    "    original_paralogs = row['Paralogs names']\n",
    "\n",
    "    # Main gene with each paralog\n",
    "    for p in paralogs:\n",
    "        pair = tuple(sorted([main_gene, p]))\n",
    "        output_rows.append({'symbol_1': pair[0], 'symbol_2': pair[1], 'Paralogs names': original_paralogs})\n",
    "\n",
    "    # Pairs among paralogs\n",
    "    for p1, p2 in combinations(paralogs, 2):\n",
    "        pair = tuple(sorted([p1, p2]))\n",
    "        output_rows.append({'symbol_1': pair[0], 'symbol_2': pair[1], 'Paralogs names': original_paralogs})\n",
    "\n",
    "# Remove duplicates\n",
    "unique_pairs_df = pd.DataFrame(output_rows).drop_duplicates(subset=['symbol_1', 'symbol_2', 'Paralogs names'])\n",
    "\n",
    "#unique_pairs_df.to_csv('gene_paralog_pairs_with_original.csv', index=False)\n",
    "\n",
    "print('Done! Check gene_paralog_pairs_with_original.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5cbc058d-17e2-4a2c-bc81-e7de91343924",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol_1</th>\n",
       "      <th>symbol_2</th>\n",
       "      <th>Paralogs names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCND2</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>CCND2, CCND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCND1</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>CCND2, CCND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCND1</td>\n",
       "      <td>CCND2</td>\n",
       "      <td>CCND2, CCND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATP1B1</td>\n",
       "      <td>ATP1B3</td>\n",
       "      <td>ATP1B1, ATP1B4, ATP1B2, ATP4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATP1B3</td>\n",
       "      <td>ATP1B4</td>\n",
       "      <td>ATP1B1, ATP1B4, ATP1B2, ATP4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1215</th>\n",
       "      <td>GCNT3</td>\n",
       "      <td>GCNT6</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216</th>\n",
       "      <td>GCNT6</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>GCNT3</td>\n",
       "      <td>GCNT7</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1218</th>\n",
       "      <td>GCNT7</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1219</th>\n",
       "      <td>GCNT3</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1220 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     symbol_1 symbol_2                                   Paralogs names\n",
       "0       CCND2    CCND3                                     CCND2, CCND1\n",
       "1       CCND1    CCND3                                     CCND2, CCND1\n",
       "2       CCND1    CCND2                                     CCND2, CCND1\n",
       "3      ATP1B1   ATP1B3                    ATP1B1, ATP1B4, ATP1B2, ATP4B\n",
       "4      ATP1B3   ATP1B4                    ATP1B1, ATP1B4, ATP1B2, ATP4B\n",
       "...       ...      ...                                              ...\n",
       "1215    GCNT3    GCNT6  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1\n",
       "1216    GCNT6    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1\n",
       "1217    GCNT3    GCNT7  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1\n",
       "1218    GCNT7    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1\n",
       "1219    GCNT3    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1\n",
       "\n",
       "[1220 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_pairs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6a8f8f76-fe7e-45a0-b9a4-294d9c44f8e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "16 input query terms found no hit:\t['ADSS', 'ADSSL1', 'MYCL1', 'H2AFV', 'H2AFZ', 'HIST1H1E', 'HIST1H1T', 'HIST1H1D', 'H1F0', 'HIST1H1B'\n"
     ]
    }
   ],
   "source": [
    "# import pandas as pd\n",
    "# import mygene\n",
    "\n",
    "# df = unique_pairs_df.copy()\n",
    "\n",
    "\n",
    "# # List of unique gene symbols from your df\n",
    "# all_genes = pd.unique(df[['symbol_1', 'symbol_2']].values.ravel())\n",
    "\n",
    "# # Initialize mygene\n",
    "# mg = mygene.MyGeneInfo()\n",
    "\n",
    "# # Query multiple fields at once\n",
    "# results = mg.querymany(\n",
    "#     all_genes,\n",
    "#     scopes='symbol',\n",
    "#     fields=[\n",
    "#         'ensembl.gene',\n",
    "#         'genomic_pos.chr',\n",
    "#         'genomic_pos.start',\n",
    "#         'genomic_pos.end',\n",
    "#         'name',            # full gene name\n",
    "#         'symbol',          # gene symbol (redundant, but can help sanity-check)\n",
    "#         'type_of_gene'     # e.g. protein-coding, pseudogene, etc.\n",
    "#     ],\n",
    "#     species='human'\n",
    "# )\n",
    "\n",
    "# # Convert to DataFrame\n",
    "# annotation_df = pd.DataFrame(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ad39a395-5dd8-4e3d-bb19-c8e49e53bbee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from biomart import BiomartServer\n",
    "import io\n",
    "\n",
    "# Connect to BioMart\n",
    "server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "dataset = server.datasets['hsapiens_gene_ensembl']\n",
    "\n",
    "# Gene list from your DataFrame\n",
    "genes = list(pd.unique(df[['symbol_1', 'symbol_2']].values.ravel()))\n",
    "\n",
    "# Perform the query\n",
    "response = dataset.search({\n",
    "    'filters': {\n",
    "        'hgnc_symbol': genes\n",
    "    },\n",
    "    'attributes': [\n",
    "        'ensembl_gene_id',\n",
    "        'hgnc_symbol',\n",
    "        'chromosome_name',\n",
    "    ]\n",
    "})\n",
    "\n",
    "# ✅ response is a `requests.Response` object, so use .text\n",
    "response_text = response.text  # ← this is the fix\n",
    "\n",
    "# Parse with pandas\n",
    "annotation_df = pd.read_csv(io.StringIO(response_text), sep='\\t', header=None)\n",
    "\n",
    "# Add column names\n",
    "annotation_df.columns = [\n",
    "    'gene', 'symbol', 'chr'\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "372cbf3e-6d32-42f7-be36-4466859505c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def extract_ensembl(ens):\n",
    "#     if isinstance(ens, dict):\n",
    "#         return ens.get('gene')\n",
    "#     elif isinstance(ens, list) and len(ens) > 0 and isinstance(ens[0], dict):\n",
    "#         return ens[0].get('gene')\n",
    "#     return None\n",
    "\n",
    "# def extract_chr(pos):\n",
    "#     if isinstance(pos, dict):\n",
    "#         return pos.get('chr')\n",
    "#     elif isinstance(pos, list) and len(pos) > 0 and isinstance(pos[0], dict):\n",
    "#         return pos[0].get('chr')\n",
    "#     return None\n",
    "\n",
    "# # Apply to annotation_df\n",
    "# annotation_df['gene'] = annotation_df['ensembl'].apply(extract_ensembl)\n",
    "# annotation_df['chr'] = annotation_df['genomic_pos'].apply(extract_chr)\n",
    "\n",
    "\n",
    "# # Keep only necessary columns\n",
    "# annotation_df = annotation_df[['gene', 'chr', 'symbol']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "581cb473-decb-44fb-bdf7-e454c8647688",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol_1</th>\n",
       "      <th>symbol_2</th>\n",
       "      <th>Paralogs names</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>chr_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>chr_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCND2</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>CCND2, CCND1</td>\n",
       "      <td>ENSG00000118971</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000112576</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCND1</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>CCND2, CCND1</td>\n",
       "      <td>ENSG00000110092</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSG00000112576</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCND1</td>\n",
       "      <td>CCND2</td>\n",
       "      <td>CCND2, CCND1</td>\n",
       "      <td>ENSG00000110092</td>\n",
       "      <td>11</td>\n",
       "      <td>ENSG00000118971</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATP1B1</td>\n",
       "      <td>ATP1B3</td>\n",
       "      <td>ATP1B1, ATP1B4, ATP1B2, ATP4B</td>\n",
       "      <td>ENSG00000143153</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000069849</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATP1B3</td>\n",
       "      <td>ATP1B4</td>\n",
       "      <td>ATP1B1, ATP1B4, ATP1B2, ATP4B</td>\n",
       "      <td>ENSG00000069849</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000101892</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1609</th>\n",
       "      <td>GCNT3</td>\n",
       "      <td>GCNT7</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>ENSG00000140297</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSG00000124091</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1610</th>\n",
       "      <td>GCNT7</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>ENSG00000124091</td>\n",
       "      <td>20</td>\n",
       "      <td>ENSG00000103489</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1611</th>\n",
       "      <td>GCNT7</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>ENSG00000124091</td>\n",
       "      <td>20</td>\n",
       "      <td>ENSG00000285395</td>\n",
       "      <td>HG2263_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1612</th>\n",
       "      <td>GCNT3</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>ENSG00000140297</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSG00000103489</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1613</th>\n",
       "      <td>GCNT3</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>ENSG00000140297</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSG00000285395</td>\n",
       "      <td>HG2263_PATCH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1614 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     symbol_1 symbol_2                                   Paralogs names  \\\n",
       "0       CCND2    CCND3                                     CCND2, CCND1   \n",
       "1       CCND1    CCND3                                     CCND2, CCND1   \n",
       "2       CCND1    CCND2                                     CCND2, CCND1   \n",
       "3      ATP1B1   ATP1B3                    ATP1B1, ATP1B4, ATP1B2, ATP4B   \n",
       "4      ATP1B3   ATP1B4                    ATP1B1, ATP1B4, ATP1B2, ATP4B   \n",
       "...       ...      ...                                              ...   \n",
       "1609    GCNT3    GCNT7  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1610    GCNT7    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1611    GCNT7    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1612    GCNT3    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1613    GCNT3    XYLT1  GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "\n",
       "               gene_1 chr_1           gene_2         chr_2  \n",
       "0     ENSG00000118971    12  ENSG00000112576             6  \n",
       "1     ENSG00000110092    11  ENSG00000112576             6  \n",
       "2     ENSG00000110092    11  ENSG00000118971            12  \n",
       "3     ENSG00000143153     1  ENSG00000069849             3  \n",
       "4     ENSG00000069849     3  ENSG00000101892             X  \n",
       "...               ...   ...              ...           ...  \n",
       "1609  ENSG00000140297    15  ENSG00000124091            20  \n",
       "1610  ENSG00000124091    20  ENSG00000103489            16  \n",
       "1611  ENSG00000124091    20  ENSG00000285395  HG2263_PATCH  \n",
       "1612  ENSG00000140297    15  ENSG00000103489            16  \n",
       "1613  ENSG00000140297    15  ENSG00000285395  HG2263_PATCH  \n",
       "\n",
       "[1614 rows x 7 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge for symbol_1\n",
    "df = df.merge(\n",
    "    annotation_df.rename(columns={\n",
    "        'symbol': 'symbol_1',\n",
    "        'gene': 'gene_1',\n",
    "        'chr': 'chr_1'\n",
    "    }),\n",
    "    on='symbol_1',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# Merge for symbol_2\n",
    "df = df.merge(\n",
    "    annotation_df.rename(columns={\n",
    "        'symbol': 'symbol_2',\n",
    "        'gene': 'gene_2',\n",
    "        'chr': 'chr_2'\n",
    "    }),\n",
    "    on='symbol_2',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "430f0085-abd1-4259-8326-4aef31b3fc79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol_1</th>\n",
       "      <th>symbol_2</th>\n",
       "      <th>Paralogs names</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>chr_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>chr_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ADSS</td>\n",
       "      <td>ADSSL1</td>\n",
       "      <td>ADSSL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>MYCL1</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>MYCN, MYCL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000134323</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>H2AFV</td>\n",
       "      <td>H2AFZ</td>\n",
       "      <td>H2AFV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>871</th>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>872</th>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>873</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>874</th>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>875</th>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876</th>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>877</th>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1E</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878</th>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>879</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1T</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886</th>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>887</th>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888</th>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1D</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892</th>\n",
       "      <td>H1F0</td>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894</th>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1B</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1C</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td>H1FX</td>\n",
       "      <td>HIST1H1A</td>\n",
       "      <td>HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>MYCL1</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>MYC, MYCL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000134323</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312</th>\n",
       "      <td>CXorf23</td>\n",
       "      <td>THRAP3</td>\n",
       "      <td>CXorf23, THRAP3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000054118</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1575</th>\n",
       "      <td>GCNT6</td>\n",
       "      <td>XYLT2</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000015532</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>GCNT6</td>\n",
       "      <td>GCNT7</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000124091</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>GCNT6</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000103489</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1608</th>\n",
       "      <td>GCNT6</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285395</td>\n",
       "      <td>HG2263_PATCH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      symbol_1  symbol_2                                     Paralogs names  \\\n",
       "16        ADSS    ADSSL1                                             ADSSL1   \n",
       "101      MYCL1      MYCN                                        MYCN, MYCL1   \n",
       "114      H2AFV     H2AFZ                                              H2AFV   \n",
       "871   HIST1H1E  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "872   HIST1H1D  HIST1H1E  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "873       H1F0  HIST1H1E  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "874   HIST1H1B  HIST1H1E  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "875   HIST1H1C  HIST1H1E  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "876       H1FX  HIST1H1E  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "877   HIST1H1A  HIST1H1E  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "878   HIST1H1D  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "879       H1F0  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "880   HIST1H1B  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "881   HIST1H1C  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "882       H1FX  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "883   HIST1H1A  HIST1H1T  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "884       H1F0  HIST1H1D  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "885   HIST1H1B  HIST1H1D  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "886   HIST1H1C  HIST1H1D  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "887       H1FX  HIST1H1D  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "888   HIST1H1A  HIST1H1D  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "889       H1F0  HIST1H1B  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "890       H1F0  HIST1H1C  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "891       H1F0      H1FX  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "892       H1F0  HIST1H1A  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "893   HIST1H1B  HIST1H1C  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "894       H1FX  HIST1H1B  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "895   HIST1H1A  HIST1H1B  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "896       H1FX  HIST1H1C  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "897   HIST1H1A  HIST1H1C  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "898       H1FX  HIST1H1A  HIST1H1T, HIST1H1D, H1F0, HIST1H1B, HIST1H1C, ...   \n",
       "1181     MYCL1      MYCN                                         MYC, MYCL1   \n",
       "1312   CXorf23    THRAP3                                    CXorf23, THRAP3   \n",
       "1575     GCNT6     XYLT2    GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1605     GCNT6     GCNT7    GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1607     GCNT6     XYLT1    GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "1608     GCNT6     XYLT1    GCNT1, GCNT2, GCNT4, GCNT6, GCNT7, GCNT3, XYLT1   \n",
       "\n",
       "     gene_1 chr_1           gene_2         chr_2  \n",
       "16      NaN   NaN              NaN           NaN  \n",
       "101     NaN   NaN  ENSG00000134323             2  \n",
       "114     NaN   NaN              NaN           NaN  \n",
       "871     NaN   NaN              NaN           NaN  \n",
       "872     NaN   NaN              NaN           NaN  \n",
       "873     NaN   NaN              NaN           NaN  \n",
       "874     NaN   NaN              NaN           NaN  \n",
       "875     NaN   NaN              NaN           NaN  \n",
       "876     NaN   NaN              NaN           NaN  \n",
       "877     NaN   NaN              NaN           NaN  \n",
       "878     NaN   NaN              NaN           NaN  \n",
       "879     NaN   NaN              NaN           NaN  \n",
       "880     NaN   NaN              NaN           NaN  \n",
       "881     NaN   NaN              NaN           NaN  \n",
       "882     NaN   NaN              NaN           NaN  \n",
       "883     NaN   NaN              NaN           NaN  \n",
       "884     NaN   NaN              NaN           NaN  \n",
       "885     NaN   NaN              NaN           NaN  \n",
       "886     NaN   NaN              NaN           NaN  \n",
       "887     NaN   NaN              NaN           NaN  \n",
       "888     NaN   NaN              NaN           NaN  \n",
       "889     NaN   NaN              NaN           NaN  \n",
       "890     NaN   NaN              NaN           NaN  \n",
       "891     NaN   NaN              NaN           NaN  \n",
       "892     NaN   NaN              NaN           NaN  \n",
       "893     NaN   NaN              NaN           NaN  \n",
       "894     NaN   NaN              NaN           NaN  \n",
       "895     NaN   NaN              NaN           NaN  \n",
       "896     NaN   NaN              NaN           NaN  \n",
       "897     NaN   NaN              NaN           NaN  \n",
       "898     NaN   NaN              NaN           NaN  \n",
       "1181    NaN   NaN  ENSG00000134323             2  \n",
       "1312    NaN   NaN  ENSG00000054118             1  \n",
       "1575    NaN   NaN  ENSG00000015532            17  \n",
       "1605    NaN   NaN  ENSG00000124091            20  \n",
       "1607    NaN   NaN  ENSG00000103489            16  \n",
       "1608    NaN   NaN  ENSG00000285395  HG2263_PATCH  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.gene_1.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4a6ca92d-1f13-4619-b8d7-25a89732ae3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, symbol, chr]\n",
       "Index: []"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_df[annotation_df.symbol == 'ADSS' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "2bd9cb32-c7cd-49d6-a61b-d1c0fe26f195",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CCND2', 'CCND3', 'CCND1', 'ATP1B1', 'ATP1B3', 'ATP1B4', 'ATP1B2',\n",
       "       'ATP4B', 'DNM1', 'DNM2', 'DNM3', 'ADSS', 'ADSSL1', 'GFPT1',\n",
       "       'GFPT2', 'CDK4', 'CDK6', 'RHOA', 'RHOB', 'RHOC', 'NMNAT1',\n",
       "       'NMNAT3', 'NMNAT2', 'RCOR1', 'RCOR2', 'RCOR3', 'TRPM2', 'TRPM7',\n",
       "       'TRPM5', 'TRPM8', 'TRPM6', 'TRPM1', 'TRPM3', 'TRPM4', 'CAB39',\n",
       "       'CAB39L', 'WNK1', 'WNK4', 'EPHA6', 'FER', 'EPHA1', 'FES', 'WNK2',\n",
       "       'WNK3', 'MYC', 'MYCN', 'MYCL1', 'PLK1', 'PLK2', 'PLK5', 'PLK3',\n",
       "       'HGS', 'WDFY2', 'WDFY1', 'MYB', 'MYBL2', 'MYBL1', 'H2AFV', 'H2AFZ',\n",
       "       'FAM50A', 'FAM50B', 'MED12', 'MED12L', 'TUBB', 'TUBB4B', 'TUBB8',\n",
       "       'TUBB4A', 'TUBB2B', 'TUBB3', 'TUBB2A', 'TUBB6', 'TUBB1', 'RBM47',\n",
       "       'SYNCRIP', 'DND1', 'RBM46', 'A1CF', 'HNRNPR', 'BTRC', 'FBXW11',\n",
       "       'STAG1', 'STAG2', 'STAG3', 'GMPR', 'IMPDH2', 'GMPR2', 'IMPDH1',\n",
       "       'EEF1A1', 'EEF1A2', 'SDC1', 'SDC3', 'SDC4', 'SDC2', 'CHMP4B',\n",
       "       'CHMP4C', 'CHMP4A', 'COPZ1', 'COPZ2', 'PFN1', 'PFN2', 'PFN3',\n",
       "       'NAA10', 'NAA11', 'RAD21', 'RAD21L1', 'ATP6V1B1', 'ATP6V1B2',\n",
       "       'BCL2', 'MCL1', 'BOK', 'BCL2L1', 'BAX', 'BCL2A1', 'BAK1', 'BCL2L2',\n",
       "       'PMM1', 'PMM2', 'DPF2', 'DPF3', 'PHF10', 'DPF1', 'RBM14', 'RBM4',\n",
       "       'RBM4B', 'SMARCA1', 'SMARCA5', 'PDS5A', 'PDS5B', 'KRAS', 'RALA',\n",
       "       'RRAS', 'ERAS', 'RALB', 'NRAS', 'HRAS', 'RRAS2', 'MRAS', 'IGF2',\n",
       "       'INS', 'IGF1', 'RAB10', 'RAB8B', 'RAB12', 'RAB8A', 'RAB13',\n",
       "       'RAB15', 'WWTR1', 'YAP1', 'SCD', 'SCD5', 'ACACA', 'ACACB', 'RPL22',\n",
       "       'RPL22L1', 'SAMD4A', 'SAMD4B', 'PIK3CA', 'PIK3CD', 'PIK3C2B',\n",
       "       'PIK3C2G', 'PIK3CG', 'PIK3CB', 'PIK3C2A', 'DDX17', 'DDX5', 'PAX1',\n",
       "       'PAX8', 'PAX5', 'PAX2', 'PAX9', 'FERMT2', 'TLN2', 'FERMT3', 'TLN1',\n",
       "       'FERMT1', 'MAPK1', 'MAPK3', 'PPP2R2A', 'PPP2R2D', 'PPP2R2B',\n",
       "       'PPP2R2C', 'EGLN1', 'EGLN2', 'EGLN3', 'TERF1', 'TERF2', 'PTK2',\n",
       "       'TNK1', 'TNK2', 'PTK2B', 'DDX21', 'DDX50', 'PPA1', 'PPA2', 'NACA',\n",
       "       'NACAD', 'NACA2', 'RAC1', 'RHOQ', 'RAC2', 'CDC42', 'RHOJ', 'RAC3',\n",
       "       'RHOG', 'PPP2R1A', 'PPP2R1B', 'ATL2', 'ATL3', 'ATL1', 'TUBA1B',\n",
       "       'TUBA3C', 'TUBA4A', 'TUBA1C', 'TUBA3D', 'TUBA1A', 'TUBAL3',\n",
       "       'TUBA3E', 'TUBA8', 'UBE2D1', 'UBE2D3', 'UBE2D4', 'UBE2D2', 'PSMB6',\n",
       "       'PSMB9', 'ILF3', 'ZFR2', 'ILF2', 'STRBP', 'ZFR', 'BRD2', 'BRD4',\n",
       "       'BRDT', 'BRD3', 'ZFX', 'ZNF639', 'ZFY', 'ZNF711', 'ASF1A', 'ASF1B',\n",
       "       'VPS4A', 'VPS4B', 'MYH11', 'MYH9', 'MYH14', 'MYH10', 'BCAR1',\n",
       "       'NEDD9', 'CASS4', 'EFS', 'NDE1', 'NDEL1', 'NCKAP1', 'NCKAP1L',\n",
       "       'EFR3A', 'EFR3B', 'RAB6A', 'RAB6C', 'RAB41', 'RAB6B', 'ARHGEF6',\n",
       "       'ARHGEF7', 'PLEKHN1', 'IRS1', 'IRS2', 'IRS4', 'IKZF3', 'IKZF4',\n",
       "       'IKZF1', 'IKZF2', 'ATP2C1', 'ATP2C2', 'MTA2', 'MTA3', 'MTA1',\n",
       "       'MBNL1', 'MBNL3', 'ZC3H10', 'MBNL2', 'CPD', 'CPZ', 'CPE', 'CPN1',\n",
       "       'CPM', 'CPXM1', 'CPXM2', 'AEBP1', 'ZNF217', 'ZNF219', 'ZNF536',\n",
       "       'ZNF516', 'EPC1', 'EPC2', 'SLC2A1', 'SLC2A5', 'SLC2A7', 'SLC2A3',\n",
       "       'SLC2A9', 'SLC2A14', 'SLC2A2', 'SLC2A4', 'SLC2A11', 'HIST1H1E',\n",
       "       'HIST1H1T', 'HIST1H1D', 'H1F0', 'HIST1H1B', 'HIST1H1C', 'H1FX',\n",
       "       'HIST1H1A', 'PRKAR1A', 'PRKAR1B', 'PRKAR2A', 'PRKAR2B', 'CHKA',\n",
       "       'CHKB', 'ETNK1', 'ETNK2', 'HK2', 'HK3', 'HKDC1', 'HK1', 'GCK',\n",
       "       'RPS4X', 'RPS4Y2', 'RPS4Y1', 'STX2', 'STX4', 'STX1A', 'STX1B',\n",
       "       'STX11', 'STX3', 'STX19', 'SP1', 'SP4', 'SP3', 'SP5', 'SP8', 'SP2',\n",
       "       'SP7', 'SP9', 'EIF5A', 'EIF5AL1', 'EIF5A2', 'SNAP23', 'SNAP25',\n",
       "       'DHX29', 'DHX36', 'DHX57', 'ESRP1', 'HNRNPH1', 'HNRNPH2', 'GRSF1',\n",
       "       'HNRNPF', 'ESRP2', 'HNRNPH3', 'ATP6V0D1', 'ATP6V0D2', 'NMT1',\n",
       "       'NMT2', 'EIF4A1', 'EIF4A2', 'RPP25', 'RPP25L', 'VRK1', 'VRK3',\n",
       "       'VRK2', 'PABPC1', 'PABPC1L', 'PABPC3', 'PABPC5', 'PABPC4',\n",
       "       'PABPC4L', 'PPARGC1B', 'PPRC1', 'PPARGC1A', 'ALDOA', 'ALDOC',\n",
       "       'ALDOB', 'RRP1', 'RRP1B', 'E2F1', 'E2F3', 'E2F2', 'E2F5', 'E2F4',\n",
       "       'E2F6', 'ANKRD11', 'BARD1', 'TONSL', 'ANKRD12', 'ANKRD31',\n",
       "       'GATAD2A', 'GATAD2B', 'FAM98B', 'FAM98C', 'FAM98A', 'DDX19A',\n",
       "       'DDX25', 'DDX19B', 'CCNT1', 'CCNT2', 'GLS', 'GLS2', 'SLC7A1',\n",
       "       'SLC7A2', 'SLC7A4', 'SLC7A14', 'SLC7A3', 'CEPT1', 'CHPT1', 'EPT1',\n",
       "       'RRAGC', 'RRAGD', 'SLC4A10', 'SLC4A7', 'SLC4A4', 'SLC4A1',\n",
       "       'SLC4A11', 'SLC4A5', 'SLC4A8', 'SLC4A2', 'SLC4A9', 'SLC4A3',\n",
       "       'TBL1XR1', 'TBL1Y', 'WDR17', 'TBL1X', 'SLC7A5', 'SLC7A7', 'SLC7A6',\n",
       "       'SLC7A11', 'SLC7A8', 'SLC7A9', 'SLC7A10', 'SLC7A13', 'CCNL1',\n",
       "       'CCNL2', 'NXT1', 'NXT2', 'RBBP4', 'RBBP7', 'HSPA1L', 'HSPA8',\n",
       "       'HSPA1B', 'HSPA6', 'HSPA2', 'HSPA1A', 'TRA2A', 'TRA2B', 'BCLAF1',\n",
       "       'CXorf23', 'THRAP3', 'SMARCA2', 'SMARCA4', 'CDH2', 'DSC1', 'CDH4',\n",
       "       'CDH16', 'CDH3', 'CDH1', 'CDH26', 'DSC3', 'DSC2', 'CDH15', 'CDH17',\n",
       "       'AHCY', 'AHCYL1', 'AHCYL2', 'SPTSSA', 'SPTSSB', 'ZFP36L1',\n",
       "       'ZFP36L2', 'ZFP36', 'PTBP1', 'PTBP3', 'PTBP2', 'CAP1', 'CAP2',\n",
       "       'TIMM17A', 'TIMM17B', 'ENO1', 'ENO3', 'ENO2', 'ARF1', 'ARF3',\n",
       "       'ARF6', 'ARF4', 'ARL17B', 'ARF5', 'PPP1CA', 'PPP1CB', 'PPP1CC',\n",
       "       'AGGF1', 'RBM10', 'RBM6', 'RBM5', 'POC1A', 'POC1B', 'GRAP', 'GRB2',\n",
       "       'GRAPL', 'SLA', 'SLA2', 'GRAP2', 'CYB5A', 'CYB5B', 'HNRNPM',\n",
       "       'MYEF2', 'SNRPA', 'SNRPB2', 'TRAF1', 'TRAF2', 'TRAF4', 'TRAF5',\n",
       "       'TRAF3', 'TRAF6', 'PPP2CA', 'PPP4C', 'PPP2CB', 'ACTA2', 'ACTB',\n",
       "       'POTEF', 'ACTG2', 'POTEI', 'POTEJ', 'ACTRT2', 'ACTC1', 'ACTRT1',\n",
       "       'ACTG1', 'ACTL7A', 'ACTBL2', 'ACTL7B', 'ACTA1', 'MAGOH', 'MAGOHB',\n",
       "       'GCNT1', 'XYLT2', 'GCNT2', 'GCNT4', 'GCNT6', 'GCNT7', 'GCNT3',\n",
       "       'XYLT1'], dtype=object)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ADSS ENSG00000185100  chr 1 \n",
    "\n",
    "ADSSL1 ENSG00000185100 chr 14 \n",
    "\n",
    "GCNT6 ENSG00000205318 chr 6 \n",
    "XYLT1 ENSG00000103489  chr 16 \n",
    "\n",
    "\n",
    "GCNT7 ENSG00000124091 chr 20\n",
    "\n",
    "XYLT2\n",
    "\n",
    "CXorf23\t\n",
    "THRAP3\n",
    "\n",
    "MYCL1\t\n",
    "\n",
    "MYCN\n",
    "\n",
    "\n",
    "H1FX\n",
    "\n",
    "\n",
    "HIST1H1A\n",
    "HIST1H1B\n",
    "HIST1H1C\n",
    "HIST1H1D\n",
    "H1F0\n",
    "\n",
    "\n",
    "H2AFV\t\n",
    "\n",
    "H2AFZ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9c8523b-8ad5-4a05-92e2-54045481f50f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
